### ISSUE TITLE
Python 3.13 support for PyTorch

### ISSUE THREAD

**Issue by rgommers** [pytorch/pytorch#130249]
### Update Jan 13 2025:

Now that enablement in trunk has been done and 3.13/3.13t are officially released, revamping this to be a support matrix for what works today and what we plan on adding:

Source build 
| OS | 3.13 | 3.13t |
| -- | -- | -- |
| Linux | :white_check_mark:  | :white_check_mark: |
| MacOS | :white_check_mark: | :white_check_mark: |
| Windows | :white_check_mark: | :white_check_mark: |

Nightlies (As of Jan 13 2025):
| OS | 3.13 | 3.13t |
| -- | -- | -- |
| Linux | :white_check_mark:  | :white_check_mark: |
| Linux Aarch64 | :white_check_mark:   | :x: |
| MacOS | :white_check_mark:  | :x: |
| Windows | :white_check_mark:  | :x: |

Release 2.5.0
| OS | 3.13 | 3.13t |
| -- | -- | -- |
| Linux | :white_check_mark:  | :x: |
| Linux Aarch64 | :x:   | :x: |
| MacOS | :x: | :x: |
| Windows | :x: | :x: |

Release 2.5.1 
| OS | 3.13 | 3.13t |
| -- | -- | -- |
| Linux | :white_check_mark:  | :x: |
| Linux Aarch64 | :x:   | :x: |
| MacOS | :x: | :x: |
| Windows | :x: | :x: |

Release 2.6 
| OS | 3.13 | 3.13t |
| -- | -- | -- |
| Linux | :white_check_mark:  | :white_check_mark: |
| Linux Aarch64 | :white_check_mark:  | :x: |
| MacOS | :white_check_mark: | :x: |
| Windows | :white_check_mark: | :x: |

Release 2.7 (tentative)
| OS | 3.13 | 3.13t |
| -- | -- | -- |
| Linux | :white_check_mark:  | :white_check_mark: |
| Linux Aarch64 | :white_check_mark:  | :white_check_mark: |
| MacOS | :white_check_mark: | :white_check_mark: |
| Windows | :white_check_mark: | :white_check_mark: |

### Original post

Python 3.13 is coming out in the near future - RC1 is planned for July 30 (first point at which the ABI is stable and packages can start releasing wheels for 3.13), and the fina release is planned for 1 Oct 2024 (see [PEP 719](https://peps.python.org/pep-0719/)). 

I see that work on supporting 3.13 has started (gh-121979, gh-130030; Cc @atalman), but I couldn't find a tracking issue for the current status of it, so I thought it'd be useful to have one. The other motivation to open an issue now: we're working on implementing support for free-threaded CPython (a.k.a. "no-GIL", [PEP 703](https://peps.python.org/pep-0703/)) and are interested to start work on a first package that depends on PyTorch. Support for the default CPython 3.13 is a pre-requisite for starting that work though.

Here is a hopefully useful breakdown in steps to support Python 3.13:
- [x] PyTorch builds from source on 3.13 (done in https://github.com/pytorch/pytorch/pull/126033 )
- [x] Nightly `cp313` wheels available for all platforms (waiting until final cpython rc)
- [x] `torch.compile` support for 3.13 (complete @williamwen42 )
- [x] PyTorch builds from source with free-threaded CPython (blocked on dependency support + update in core, tracking PR https://github.com/pytorch/pytorch/pull/130689)
- [x] `cp313t` (free-threaded) nightly wheels available (blocked on ft build and dependencies)

Python 3.13 CD:

- [x] Nightly `cp313` Linux wheels available: https://github.com/pytorch/pytorch/pull/130030 
- [x] https://github.com/pytorch/builder/pull/2035 
- [x] Nightly windows wheels : https://github.com/pytorch/pytorch/pull/138095
- [x] Nightly macos: https://github.com/pytorch/pytorch/pull/140137
- [x] Nightly aarch64 wheels : https://github.com/pytorch/pytorch/pull/138629
- [x] https://github.com/pytorch/pytorch/pull/136001
- [x] https://github.com/pytorch/pytorch/pull/137127
- [x] https://github.com/pytorch/builder/pull/2001
- [x] https://github.com/pytorch/pytorch/pull/142294

Python 3.13 CI & torch.compile:

- [x] https://github.com/pytorch/pytorch/pull/143162
- [x] https://github.com/pytorch/pytorch/pull/143078 
- [x] https://github.com/pytorch/pytorch/pull/143074
- [x] #141264
- [x] #141951
- [x] #141950
- [x] #141887
- [x] #141886
- [x] #141860
- [x] #141859
- [x] #140733
- [x] #139533
- [x] #141862
- [x] #141858
- [x] #141674
- [x] #141673
- [x] #141623
- [x] #141621
- [x] #141605
- [x] #141577
- [x] #141572
- [x] #142003
- [x] #141409

`torch.compile` is broken out separately based on the history for Python 3.12 (see gh-110436 for general support, and gh-120233 for `torch.compile`).

The status of dependencies should be fine AFAIK. E.g. Pybind11 2.13.1 has full support, and SymPy `1.13.0rc4` has too (xref [sympy#26692](https://github.com/sympy/sympy/issues/26692)). 

For free-threaded support we should probably open a separate tracking issue once regular 3.13 support is in good shape. I'll note here that Pybind11 and NumPy already have mostly-complete support for it, and NumPy should release (experimental) wheels shortly after Python 3.13.0rc1).

cc @malfet @seemethere @pytorch/pytorch-dev-infra @albanD

**Comment by albanD** [comment]
Status of 3.13t testing.

This status is accurate as of 07/17 (with a CPython from the 07/12  to avoid https://github.com/numpy/numpy/issues/26157#issuecomment-2234330798) , based on https://github.com/pytorch/pytorch/pull/130689 with dependencies installed with `pip install --pre -i https://pypi.anaconda.org/scientific-python-nightly-wheels/simple -r .ci/docker/requirements-ci.txt ` (the version of this file from the PR).

Running `python run_test.py --core` in with all 3 versions of (3.13, 3.13t with PYTHON_GIL=1 and 3.13t with PYTHON_GIL=0).

Only `TestFwdGradientsCPU.test_forward_mode_AD_linalg_det_singular_cpu_float64` is failing.
Every other core test is green.

<details>
  <summary>Full Error</summary>
```
______________________ TestFwdGradientsCPU.test_forward_mode_AD_linalg_det_singular_cpu_float64 ______________________
Traceback (most recent call last):
  File "/home/albandes/local/pytorch/3.13.0a6_debug_source/torch/testing/_internal/common_device_type.py", line 1127, in test_wrapper
    return test(*args, **kwargs)
  File "/home/albandes/local/pytorch/3.13.0a6_debug_source/test/test_ops_fwd_gradients.py", line 90, in test_forward_mode_AD
    self._forward_grad_helper(device, dtype, op, op.get_op(), is_inplace=False)
    ~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/albandes/local/pytorch/3.13.0a6_debug_source/test/test_ops_fwd_gradients.py", line 72, in _forward_grad_helper
    call_grad_test_helper()
    ~~~~~~~~~~~~~~~~~~~~~^^
  File "/home/albandes/local/pytorch/3.13.0a6_debug_source/test/test_ops_fwd_gradients.py", line 60, in call_grad_test_helper
    self._grad_test_helper(
    ~~~~~~~~~~~~~~~~~~~~~~^
        device,
        ^^^^^^^
    ...<6 lines>...
        check_batched_forward_grad=check_batched_forward_grad,
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/home/albandes/local/pytorch/3.13.0a6_debug_source/torch/testing/_internal/common_utils.py", line 5021, in _grad_test_helper
    return self._check_helper(device, dtype, op, variant, 'gradcheck', check_forward_ad=check_forward_ad,
           ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                              check_backward_ad=check_backward_ad, check_batched_grad=check_batched_grad,
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                              check_batched_forward_grad=check_batched_forward_grad)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/albandes/local/pytorch/3.13.0a6_debug_source/torch/testing/_internal/common_utils.py", line 4990, in _check_helper
    self.assertTrue(gradcheck(fn, gradcheck_args,
                    ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^
                              check_batched_grad=check_batched_grad,
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<5 lines>...
                              check_undefined_grad=True,
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^
                              check_batched_forward_grad=check_batched_forward_grad))
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/albandes/local/pytorch/3.13.0a6_debug_source/torch/testing/_internal/common_utils.py", line 4560, in gradcheck
    return torch.autograd.gradcheck(fn, inputs, **kwargs)
           ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^
  File "/home/albandes/local/pytorch/3.13.0a6_debug_source/torch/autograd/gradcheck.py", line 2053, in gradcheck
    return _gradcheck_helper(**args)
  File "/home/albandes/local/pytorch/3.13.0a6_debug_source/torch/autograd/gradcheck.py", line 2082, in _gradcheck_helper
    _gradcheck_real_imag(
    ~~~~~~~~~~~~~~~~~~~~^
        gradcheck_fn,
        ^^^^^^^^^^^^^
    ...<11 lines>...
        check_undefined_grad=check_undefined_grad,
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/home/albandes/local/pytorch/3.13.0a6_debug_source/torch/autograd/gradcheck.py", line 1559, in _gradcheck_real_imag
    gradcheck_fn(
    ~~~~~~~~~~~~^
        func,
        ^^^^^
    ...<8 lines>...
        use_forward_ad=True,
        ^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/home/albandes/local/pytorch/3.13.0a6_debug_source/torch/autograd/gradcheck.py", line 1926, in _fast_gradcheck
    _check_analytical_numerical_equal(
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
        analytical_vJu,
        ^^^^^^^^^^^^^^^
    ...<11 lines>...
        is_forward_ad=use_forward_ad,
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/home/albandes/local/pytorch/3.13.0a6_debug_source/torch/autograd/gradcheck.py", line 1855, in _check_analytical_numerical_equal
    raise GradcheckError(
    ...<4 lines>...
    )
torch.autograd.gradcheck.GradcheckError: Jacobian computed with forward mode mismatch for output 0 with respect to input 0,
numerical:tensor([-0.0821,  0.0224], dtype=torch.float64)
analytical:tensor([-0.0821,  0.0000], dtype=torch.float64)

The above quantities relating the numerical and analytical jacobians are computed 
in fast mode. See: https://github.com/pytorch/pytorch/issues/53876 for more background 
about fast mode. Below, we recompute numerical and analytical jacobians in slow mode:

Numerical:
 tensor([[-0.0936,  0.0000],
        [-0.1266,  0.0000],
        [ 0.0897,  0.0000],
        [-0.1096,  0.0000],
        [-0.0974,  0.0000],
        [-0.0908,  0.0000],
        [-0.1228,  0.0000],
        [ 0.0870,  0.0000],
        [-0.1063,  0.0000],
        [-0.0945,  0.0000],
        [ 0.0300,  0.0000],
        [ 0.0406,  0.0000],
        [-0.0287,  0.0000],
        [ 0.0351,  0.0000],
        [ 0.0312,  0.0000],
        [-0.0355,  0.0000],
        [-0.0480,  0.0000],
        [ 0.0340,  0.0000],
        [-0.0416,  0.0000],
        [-0.0369,  0.0000],
        [ 0.0208,  0.0000],
        [ 0.0282,  0.0000],
        [-0.0200,  0.0000],
        [ 0.0244,  0.0000],
        [ 0.0217,  0.0000],
        [ 0.0000,  0.0793],
        [ 0.0000, -0.0102],
        [ 0.0000, -0.1129],
        [ 0.0000,  0.0684],
        [ 0.0000, -0.0278],
        [ 0.0000, -0.0177],
        [ 0.0000,  0.0023],
        [ 0.0000,  0.0251],
        [ 0.0000, -0.0152],
        [ 0.0000,  0.0062],
        [ 0.0000,  0.0344],
        [ 0.0000, -0.0044],
        [ 0.0000, -0.0490],
        [ 0.0000,  0.0297],
        [ 0.0000, -0.0121],
        [ 0.0000, -0.0785],
        [ 0.0000,  0.0100],
        [ 0.0000,  0.1116],
        [ 0.0000, -0.0677],
        [ 0.0000,  0.0275],
        [ 0.0000, -0.0762],
        [ 0.0000,  0.0098],
        [ 0.0000,  0.1085],
        [ 0.0000, -0.0657],
        [ 0.0000,  0.0268]], dtype=torch.float64)
Analytical:
tensor([[-0.0936,  0.0000],
        [-0.1266,  0.0000],
        [ 0.0897,  0.0000],
        [-0.1096,  0.0000],
        [-0.0974,  0.0000],
        [-0.0908,  0.0000],
        [-0.1228,  0.0000],
        [ 0.0870,  0.0000],
        [-0.1063,  0.0000],
        [-0.0945,  0.0000],
        [ 0.0300,  0.0000],
        [ 0.0406,  0.0000],
        [-0.0287,  0.0000],
        [ 0.0351,  0.0000],
        [ 0.0312,  0.0000],
        [-0.0355,  0.0000],
        [-0.0480,  0.0000],
        [ 0.0340,  0.0000],
        [-0.0416,  0.0000],
        [-0.0369,  0.0000],
        [ 0.0208,  0.0000],
        [ 0.0282,  0.0000],
        [-0.0200,  0.0000],
        [ 0.0244,  0.0000],
        [ 0.0217,  0.0000],
        [-0.0000,  0.0000],
        [-0.0000, -0.0000],
        [-0.0000, -0.0000],
        [-0.0000,  0.0000],
        [-0.0000, -0.0000],
        [-0.0000, -0.0000],
        [-0.0000,  0.0000],
        [-0.0000,  0.0000],
        [-0.0000, -0.0000],
        [-0.0000,  0.0000],
        [-0.0000,  0.0000],
        [-0.0000, -0.0000],
        [-0.0000, -0.0000],
        [-0.0000,  0.0000],
        [-0.0000, -0.0000],
        [-0.0000, -0.0000],
        [-0.0000,  0.0000],
        [-0.0000,  0.0000],
        [-0.0000, -0.0000],
        [-0.0000,  0.0000],
        [-0.0000, -0.0000],
        [-0.0000,  0.0000],
        [-0.0000,  0.0000],
        [-0.0000, -0.0000],
        [-0.0000,  0.0000]], dtype=torch.float64)

The max per-element difference (slow mode) is: 0.11286272035088239.


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/albandes/local/pytorch/3.13.0a6_debug_source/torch/testing/_internal/common_utils.py", line 2814, in wrapper
    method(*args, **kwargs)
    ~~~~~~^^^^^^^^^^^^^^^^^
  File "/home/albandes/local/pytorch/3.13.0a6_debug_source/torch/testing/_internal/common_device_type.py", line 447, in instantiated_test
    result = test(self, **param_kwargs)
  File "/home/albandes/local/pytorch/3.13.0a6_debug_source/torch/testing/_internal/common_device_type.py", line 1197, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/home/albandes/local/pytorch/3.13.0a6_debug_source/torch/testing/_internal/common_utils.py", line 1428, in wrapper
    fn(*args, **kwargs)
    ~~^^^^^^^^^^^^^^^^^
  File "/home/albandes/local/pytorch/3.13.0a6_debug_source/torch/testing/_internal/common_device_type.py", line 1139, in test_wrapper
    raise e_tracked from e
Exception: Caused by sample input at index 19: SampleInput(input=Tensor[size=(2, 5, 5), device="cpu", dtype=torch.float64], args=(), kwargs={}, broadcasts_input=False, name='')

To execute this test, run the following from the base repo dir:
    PYTORCH_OPINFO_SAMPLE_INPUT_INDEX=19 python test/test_ops_fwd_gradients.py -k TestFwdGradientsCPU.test_forward_mode_AD_linalg_det_singular_cpu_float64

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
```

</details>

**Comment by adamjstewart** [comment]
Looks like 3.13 didn't sneak into 2.5.1, looking forward to PyPI wheels on all platforms so we can test TorchGeo with 3.13.

**Comment by pannous** [comment]
since python 3.13 is now the current (home)brew install, can we summarize a method to install torch for mainstream?

**Comment by adamjstewart** [comment]
PyTorch should work for 3.13 when built from source, there just aren't public wheels on PyPI yet.

**Comment by liudonghua123** [comment]
I tried to install pytorch via `https://pypi.org/project/torch/#files` in my python 3.13.0 on windows, but it failed with `ERROR: Could not find a version that satisfies the requirement torch (from versions: none)`.

I also checked the files on [pypi](https://pypi.org/project/torch/#files), only whl files for python 3.9-12 for windows are available.

**Comment by zhangyx1998** [comment]
Same issue here on MacOS:

```sh
$ python3 --version; pip3 --version
  Python 3.13.0
  pip 24.3.1 from /Users/***/Library/Python/3.13/lib/python/site-packages/pip (python 3.13)
```

```sh
$ pip3 install torch
  ERROR: Could not find a version that satisfies the requirement torch (from versions: none)
  ERROR: No matching distribution found for torch
```

```sh
$ pip3 install torchvision
  ERROR: Ignored the following yanked versions: 0.1.6, 0.1.7, 0.1.8, 0.1.9, 0.2.0, 0.2.1, 0.2.2, 0.2.2.post2, 0.2.2.post3
  ERROR: Could not find a version that satisfies the requirement torchvision (from versions: none)
  ERROR: No matching distribution found for torchvision
```

```sh
$ pip3 install torchaudio
  ERROR: Could not find a version that satisfies the requirement torchaudio (from versions: none)
  ERROR: No matching distribution found for torchaudio
```

**Comment by atalman** [comment]
Hi @adamjstewart @zhangyx1998 Python 3.13 nightly binaries are available on nightly. Here is the log on MacOS M1 machine:

```
pip3 install torch --index-url https://download.pytorch.org/whl/nightly/cpu
Looking in indexes: https://download.pytorch.org/whl/nightly/cpu
Collecting torch
  Downloading https://download.pytorch.org/whl/nightly/cpu/torch-2.6.0.dev20241104-cp313-none-macosx_11_0_arm64.whl (64.2 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.2/64.2 MB 24.8 MB/s eta 0:00:00
Collecting filelock (from torch)
  Downloading https://download.pytorch.org/whl/nightly/filelock-3.16.1-py3-none-any.whl (16 kB)
Collecting typing-extensions>=4.10.0 (from torch)
  Using cached https://download.pytorch.org/whl/nightly/typing_extensions-4.12.2-py3-none-any.whl (37 kB)
Collecting networkx (from torch)
  Downloading https://download.pytorch.org/whl/nightly/networkx-3.4.2-py3-none-any.whl (1.7 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 34.2 MB/s eta 0:00:00
Collecting jinja2 (from torch)
  Downloading https://download.pytorch.org/whl/nightly/jinja2-3.1.4-py3-none-any.whl (133 kB)
Collecting fsspec (from torch)
  Downloading https://download.pytorch.org/whl/nightly/fsspec-2024.10.0-py3-none-any.whl (179 kB)
Requirement already satisfied: setuptools in /Users/atalman/miniconda3/envs/py313/lib/python3.13/site-packages (from torch) (75.1.0)
Collecting sympy==1.13.1 (from torch)
  Using cached https://download.pytorch.org/whl/nightly/sympy-1.13.1-py3-none-any.whl (6.2 MB)
Collecting mpmath<1.4,>=1.1.0 (from sympy==1.13.1->torch)
  Using cached https://download.pytorch.org/whl/nightly/mpmath-1.3.0-py3-none-any.whl (536 kB)
Collecting MarkupSafe>=2.0 (from jinja2->torch)
  Downloading https://download.pytorch.org/whl/nightly/MarkupSafe-2.1.5.tar.gz (19 kB)
  Preparing metadata (setup.py) ... done
Building wheels for collected packages: MarkupSafe
  Building wheel for MarkupSafe (setup.py) ... done
  Created wheel for MarkupSafe: filename=MarkupSafe-2.1.5-cp313-cp313-macosx_11_0_arm64.whl size=14035 sha256=51bc8b91d1a7a1638818c4f02022c1373060524763ebc6626ded3384e2b03240
  Stored in directory: /Users/atalman/Library/Caches/pip/wheels/3a/21/87/28c44597225fd0c28d6ffa365f1c2c9dd0ab763711aa4957c6
Successfully built MarkupSafe
Installing collected packages: mpmath, typing-extensions, sympy, networkx, MarkupSafe, fsspec, filelock, jinja2, torch
Successfully installed MarkupSafe-2.1.5 filelock-3.16.1 fsspec-2024.10.0 jinja2-3.1.4 mpmath-1.3.0 networkx-3.4.2 sympy-1.13.1 torch-2.6.0.dev20241104 typing-extensions-4.12.2
```

**Comment by w0o** [comment]
This is breaking multiple high-profile packages out there including [Docling](https://github.com/DS4SD/docling) and [InstructLab](https://github.com/instructlab/instructlab).
The workaround to use nightly build is okay but do we have any idea of when a public (2.5.1) will be available?

**Comment by williamwen42** [comment]
Note: `torch.compile` is currently NOT supported for 3.13t.

**Comment by FerLuisxd** [comment]
Any updates on this?

**Comment by adamjstewart** [comment]
@JasonMendoza2008 where do you see this? I don't see anything in PyTorch using `collections.Iterable`.

**Comment by albanD** [comment]
@FerLuisxd something specific you're looking for?

@JasonMendoza2008 this is not needed for PyTorch to work as far as I can see. Can you give details in which context this is blocking you?

**Comment by malfet** [comment]
I'm curious if we should add a separate issue for 3.13t, but 3.13 support seems to be enabled on all the platforms, isn't it? (In the upcoming 2.6.0 release)

**Comment by JasonMendoza2008** [comment]
@adamjstewart @albanD Looks like running `pip3 install torch torchvision torchaudio` had installed an outdated version (that used `collections.Iterable`) of `torchvision` which is why I was encountering the problem. My bad.

**Comment by bittnt** [comment]
Hi, it looks like the pip install torch is still broken with python 3.13.1 on mac but downgrade to python 3.12 works fine. If this can be stated in the main doc would be appreciated.

**Comment by atalman** [comment]
Hi @bittnt torch for python 3.13.1 is available on nightly build using this command:
```
pip3 install torch --pre torch --index-url https://download.pytorch.org/whl/nightly/cpu
```
After Release 2.6 is published it should be available via pypi install: ```pip3 install torch```

Here are the timeline for release 2.6: https://dev-discuss.pytorch.org/t/pytorch-release-2-6-0-call-for-features/2574

**Comment by rgommers** [comment]
@atalman according to the tables in the issue description, `cp313t` wheels for free-threading for Linux should be available for 2.6.0, however they're missing from PyPI (while they are available on https://download.pytorch.org/whl/torch/). Is this an oversight, or on purpose?

**Comment by atalman** [comment]
hi @rgommers This was done on purpose to allow people to consume 3.13t via download.pytorch.org while these builds are still experimental, we decided not to publish them to PyPI

**Comment by rinarakaki** [comment]
@atalman 
I'm using uv. Could you tell me how exactly I can install pytorch from download.pytorch.org?

.python-version

```
3.13t
```

pyproject.toml

```toml
[project]
name = "app"
requires-python = ">=3.13"
dependencies = [
  "torch+cpu>=2.6,<3",
  "torchaudio>=2.6,<3",
]

[tool.uv.sources]
torch = [
  { index = "pytorch-cpu" },
]

[[tool.uv.index]]
name = "pytorch-cpu"
url = "https://download.pytorch.org/whl/cpu"
explicit = true
```

Input:

```sh
uv sync
```

Output:

```
Resolved 111 packages in 6.21s
error: Distribution `torch==2.6.0 @ registry+https://pypi.org/simple` can't be installed because it doesn't have a source distribution or wheel for the current platform

hint: You're on Linux (`manylinux_2_36_aarch64`), but `torch` (v2.6.0) only has wheels for the following platform: `macosx_11_0_arm64`
```

**Comment by sohang3112** [comment]
>I'm using uv. Could you tell me how exactly I can install pytorch from download.pytorch.org?

@rinarakaki Pip allows installing directly from github with this syntax in *requirements.txt* : `git+https://github.com/owner/repo` (check [this Stack Overflow answer](https://stackoverflow.com/a/35998253/12947681)) - not sure if uv supports this or not.

**Comment by rgommers** [comment]
Installing from GitHub isn't the right answer. `uv` has documented exactly how to use the PyTorch index servers: https://docs.astral.sh/uv/guides/integration/pytorch/#using-a-pytorch-index

**Comment by uvikas** [comment]
Any update on timeline of `torch.compile` support on python 3.13t no-gil?

**Comment by williamwen42** [comment]
@uvikas It is in progress and it is expected to be in PyTorch 2.8 (nightlies will support it sooner but no timeline guarantees).

**Comment by malfet** [comment]
It feels like this issue were mostly about enabling 3.13 CI/CD and 3.13t CD. Q: should we keep working on this one to add no-gil testing to CI? Or track it in separate issue?

Tentatively removing `module: build` and adding `module:ci` and milestoning for 2.9

**Comment by williamwen42** [comment]
Status of 3.13t compile support: fixes to dynamo for 3.13t have already been merged and tested locally, but there is still no 3.13t CI, so we cannot fully verify that dynamo has no issues with 3.13t (in particular, the dynamo-wrapped shard is yet to be run).

Triton does not appear to support 3.13t at the moment, so having a triton installation will result in segfaults with backend="inductor". CPU inductor appears to work without a triton installation.

**Comment by atalman** [comment]
Issue adding python 3.13t support in CI created: https://github.com/pytorch/pytorch/issues/156854
Closing General Python 3.13 support issue

**Comment by subinkoovery** [comment]
> I have Mac OS 15.4.1 and Python 3.13.0. I get the following error when try to install Pytorch pip install torch : ERROR: Could not find a version that satisfies the requirement torch (from versions: none) ERROR: No matching distribution found for torch According to the documentation, it should work with pip install torch. Please tell me what I need to do. Is there a particular URL I should be referencing in the install? Thank you.

How did you fix this, I am also facing the same. Any other solution other than rolling back to older versions.

**Comment by rweintr-t** [comment]
No. The only way I got it to work was using Python 3.12.

**Comment by leder11011** [comment]
> No. The only way I got it to work was using Python 3.12.

I don't think so: I got the following versions installed
```
   ~/Py/Pannous/projekte    main *1 !3 ?6  python --version                                                      ✔    joplin-rag 
Python 3.13.7
```

```
torch                                    2.9.0
torchvision                              0.24.0
```


@subinkoovery 
Try updating all python modules with:
```
pip-review --auto
```
